File: src/claude_genmon/local_usage.py
from __future__ import annotations

import json
import os

from claude_genmon.core import Usage, UsageStats

DEFAULT_STATS_CACHE_PATH = os.path.expanduser("~/.claude/stats-cache.json")

TOKEN_FIELDS = (
    "inputTokens",
    "outputTokens",
    "cacheReadInputTokens",
    "cacheCreationInputTokens",
)


class LocalDiskUsage(Usage):
    """Reads cumulative usage stats straight from Claude Code's local
    stats cache, with no network call."""

    def __init__(self, stats_cache_path: str = DEFAULT_STATS_CACHE_PATH):
        super().__init__()
        self.stats_cache_path = stats_cache_path

    def _fetch(self) -> UsageStats | None:
        with open(self.stats_cache_path, "r", encoding="utf-8") as f:
            cache = json.load(f)

        total_tokens = sum(
            model_stats.get(field, 0)
            for model_stats in (cache.get("modelUsage") or {}).values()
            for field in TOKEN_FIELDS
        )

        return UsageStats(
            total_sessions=cache.get("totalSessions"),
            total_messages=cache.get("totalMessages"),
            total_tokens=total_tokens,
        )

File: src/local_usage.rs
use std::path::PathBuf;

use anyhow::{Context, Result};

use crate::core::{UsageSource, UsageStats};

const TOKEN_FIELDS: [&str; 4] = [
    "inputTokens",
    "outputTokens",
    "cacheReadInputTokens",
    "cacheCreationInputTokens",
];

fn default_stats_cache_path() -> PathBuf {
    let home = std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_default();
    home.join(".claude/stats-cache.json")
}

/// Reads cumulative usage stats straight from Claude Code's local stats
/// cache, with no network call.
pub struct LocalDiskUsage {
    stats_cache_path: PathBuf,
}

impl Default for LocalDiskUsage {
    fn default() -> Self {
        Self {
            stats_cache_path: default_stats_cache_path(),
        }
    }
}

impl UsageSource for LocalDiskUsage {
    fn fetch(&self) -> Result<UsageStats> {
        let content = std::fs::read_to_string(&self.stats_cache_path)
            .with_context(|| format!("reading {}", self.stats_cache_path.display()))?;
        let cache: serde_json::Value = serde_json::from_str(&content)?;

        let total_tokens: u64 = cache
            .get("modelUsage")
            .and_then(|v| v.as_object())
            .map(|model_usage| {
                model_usage
                    .values()
                    .map(|model_stats| {
                        TOKEN_FIELDS
                            .iter()
                            .map(|field| {
                                model_stats
                                    .get(*field)
                                    .and_then(|v| v.as_u64())
                                    .unwrap_or(0)
                            })
                            .sum::<u64>()
                    })
                    .sum()
            })
            .unwrap_or(0);

        Ok(UsageStats {
            total_sessions: cache.get("totalSessions").and_then(|v| v.as_u64()),
            total_messages: cache.get("totalMessages").and_then(|v| v.as_u64()),
            total_tokens: Some(total_tokens),
            ..Default::default()
        })
    }
}
